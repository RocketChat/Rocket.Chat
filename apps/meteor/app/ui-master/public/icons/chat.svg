<svg id="chat" viewBox="0 0 20 20">
  <path d="M3.36312881,15.7593532 C3.32888874,15.7771726 3.29959994,15.7925621 3.27370934,15.8064174 C3.73031444,16.1917386 4.92393678,16.4549204 7.10274479,16.4476544 L7.21881505,16.4476539 C8.55574002,16.4476543 9.6998097,16.021859 10.6510241,15.170268 L12.188241,15.2712054 C11.5513943,15.8887667 11.2142468,16.2130071 11.1767984,16.2439265 C10.1491488,17.09241 8.76567843,17.5637907 7.21881505,17.5637907 L7.10806092,17.5644215 L6.97721254,17.5644231 C4.30241208,17.5644231 2.62246666,17.1951018 2.10341288,16.0930165 C1.82161029,15.4951149 2.13151102,15.1455492 2.84114675,14.7722012 C3.31318797,14.5227311 3.32172914,14.5103228 3.24657386,14.3473099 C3.20302404,14.2436958 3.16526615,14.1604157 3.07972552,13.9763174 C2.74839464,13.2684751 2.62592155,12.9023078 2.62592155,12.4061039 C2.62592155,11.562445 2.83045155,10.7486059 3.21214892,10.0211892 C3.42204019,9.66099218 3.78721597,9.06536542 4.270822,8.82818513 C4.270822,8.80275057 4.53792399,10.0211892 4.53792399,10 C4.45304843,10.1183326 4.36401443,10.2496015 4.270822,10.3938065 C3.78721597,11.0991197 3.7445882,11.6631394 3.7445882,12.4061039 C3.7445882,12.6985231 3.82924269,12.9394008 4.09330148,13.5047226 C4.19543836,13.7251753 4.22995648,13.801243 4.270822,13.8994264 C4.58782821,14.5810467 4.36137367,15.1036933 3.78721597,15.5080098 C3.65916127,15.5981847 3.55361796,15.6588423 3.36312881,15.7593532 Z M16.5696621,12.3226926 C16.5350205,12.3039671 16.4951623,12.2829692 16.4468239,12.2578135 C16.2265134,12.1415693 16.1048536,12.0716493 15.9574055,11.9678178 C15.2986909,11.5039571 15.0399564,10.9068098 15.4049351,10.1218159 C15.4504792,10.0121781 15.4903754,9.92425877 15.608285,9.66976084 C15.914896,9.01333804 16.0132801,8.73339397 16.0132801,8.39198996 C16.0132801,5.80517045 13.90418,3.70028722 11.3120655,3.70028722 C8.71995111,3.70028722 6.61085101,5.80517045 6.61085101,8.39198996 C6.61085101,11.1096858 8.86296092,13.082961 11.9786225,13.082961 L12.112848,13.0829616 C14.6633683,13.0914671 16.0501708,12.7797435 16.5696621,12.3226926 Z M16.5612822,10.6280948 C16.4685253,10.8294269 16.4833349,10.8509584 17.0358533,11.1429611 C17.8499187,11.5712649 18.2032866,11.969896 17.8824254,12.6506683 C17.2857779,13.9175051 15.3458564,14.3439799 12.2581228,14.3439799 L12.1073975,14.3439799 L11.9780624,14.3432467 C8.14202747,14.3432482 5.34836876,11.8401459 5.34836876,8.39198996 C5.34836876,5.10955211 8.0234498,2.44 11.3120655,2.44 C14.6008168,2.44 17.276494,5.10968724 17.276494,8.39198996 C17.276494,8.96303774 17.1353362,9.38506819 16.7528285,10.2022436 C16.6539056,10.4151439 16.6101724,10.5116034 16.5612822,10.6280948 Z"/>
</svg>