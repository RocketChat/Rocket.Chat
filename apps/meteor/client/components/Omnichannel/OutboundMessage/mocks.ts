import type { IOutboundProvider, IOutboundProviderMetadata } from '@rocket.chat/core-typings';

export const getProvidersMock = async (): Promise<IOutboundProvider[]> => {
	return [
		{
			providerId: 'providerramdonlygeneratedid',
			providerName: 'Mock WhatsApp App',
			providerType: 'phone',
			supportsTemplates: true,
		},
	];
};

export const getProviderMetadataMock = (): IOutboundProviderMetadata => ({
	providerId: 'providerramdonlygeneratedid',
	providerName: 'Mock WhatsApp App',
	providerType: 'phone',
	supportsTemplates: true,
	templates: {
		'+5547998461115': [
			{
				id: 'phonenumberramdomlygenerateid',
				name: 'Dentist Appointment',
				language: 'en',
				type: 'whatsapp',
				category: 'MARKETING',
				status: 'APPROVED',
				qualityScore: {
					score: 'GREEN',
					reasons: null,
				},
				components: [
					{
						type: 'HEADER',
						format: 'TEXT',
						text: 'New {{1}} appointment',
					},
					{
						type: 'BODY',
						text: '**Hi {{1}}** Your _appointment_ for {{2}} is scheduled for {{3}} and can be rescheduled to {{4}} if {{5}} becomes available. {{6}} what do you choose?',
					},
					{
						type: 'FOOTER',
						text: 'Need to reschedule? Tap below to reply',
					},
				],
				createdAt: '', // ISO 8601 timestamp
				createdBy: '',
				modifiedAt: '', // ISO 8601 timestamp
				modifiedBy: '',
				namespace: '',
				wabaAccountId: '',
				// This is the phone number that will be used to send the message.
				phoneNumber: '+5547998461115',
				partnerId: '',
				externalId: '',
				updatedExternal: '', // ISO 8601 timestamp
				rejectedReason: null,
			},
			{
				id: 'phonenumberramdomlygenerateid2',
				name: `Doctor's Appointment`,
				language: 'en',
				type: 'whatsapp',
				category: 'MARKETING',
				status: 'APPROVED',
				qualityScore: {
					score: 'GREEN',
					reasons: null,
				},
				components: [
					{
						type: 'HEADER',
						format: 'TEXT',
						text: 'New {{1}} appointment',
					},
					{
						type: 'BODY',
						text: '**Hi {{1}}** Your _appointment_ for {{2}} is scheduled for {{3}}',
					},
					{
						type: 'FOOTER',
						text: 'Need to reschedule? Tap below to reply',
					},
				],
				createdAt: '', // ISO 8601 timestamp
				createdBy: '',
				modifiedAt: '', // ISO 8601 timestamp
				modifiedBy: '',
				namespace: '',
				wabaAccountId: '',
				// This is the phone number that will be used to send the message.
				phoneNumber: '+5547998461115',
				partnerId: '',
				externalId: '',
				updatedExternal: '', // ISO 8601 timestamp
				rejectedReason: null,
			},
		],
	},
});
