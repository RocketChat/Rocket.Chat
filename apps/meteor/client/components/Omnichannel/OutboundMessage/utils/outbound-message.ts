import type {
	IOutboundMessage,
	IOutboundProviderTemplate,
	TemplateComponent,
	TemplateParameter as CoreTemplateParameter,
} from '@rocket.chat/core-typings';

import type { SubmitPayload } from '../components/OutboundMessageWizard/forms';
import type { MessageFormSubmitPayload } from '../components/OutboundMessageWizard/forms/MessageForm';
import type { RecipientFormSubmitPayload } from '../components/OutboundMessageWizard/forms/RecipientForm';
import type { TemplateParameter, TemplateParameters } from '../definitions/template';

export const isRecipientStepValid = (data: Partial<SubmitPayload>): data is Required<RecipientFormSubmitPayload> => {
	return !!(data.contactId && data.contact && data.provider && data.providerId && data.recipient && data.sender);
};

export const isMessageStepValid = (data: Partial<SubmitPayload>): data is Required<MessageFormSubmitPayload> => {
	return !!(data.templateId && data.template && data.templateParameters);
};

const formatParameterForOutboundMessage = (parameter: TemplateParameter): CoreTemplateParameter => {
	switch (parameter.type) {
		case 'media':
			return { type: 'media', link: parameter.value, format: parameter.format };
		default:
			return { type: 'text', text: parameter.value };
	}
};

const formatOutboundMessageComponents = (
	components: IOutboundProviderTemplate['components'],
	parameters: TemplateParameters,
): TemplateComponent[] => {
	return components.map((component) => {
		const values = parameters[component.type] || [];
		return {
			type: component.type,
			parameters: values.map(formatParameterForOutboundMessage),
		};
	});
};

export const formatOutboundMessagePayload = ({
	recipient,
	sender,
	template,
	type = 'template',
	templateParameters,
}: {
	type: IOutboundMessage['type'];
	recipient: string;
	sender: string;
	template: IOutboundProviderTemplate;
	templateParameters: TemplateParameters;
}): IOutboundMessage => {
	return {
		to: recipient,
		type,
		templateProviderPhoneNumber: sender,
		template: {
			name: template.name,
			language: {
				code: template.language,
			},
			components: formatOutboundMessageComponents(template.components, templateParameters),
		},
	};
};
