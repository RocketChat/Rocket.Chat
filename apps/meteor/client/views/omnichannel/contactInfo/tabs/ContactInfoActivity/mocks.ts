import { faker } from '@faker-js/faker';
import type { IOutboundProviderTemplate } from '@rocket.chat/core-typings';

import type { ActivityItem } from './useActivityList';

export const mockActivityList: ActivityItem[] = [
	{
		id: '123',
		name: 'Haylie George',
		status: { id: 'sent', ts: new Date().toISOString() },
	},
	{
		id: '234',
		name: 'Haylie George',
		status: { id: 'delivered', ts: new Date().toISOString() },
	},
];

const mockTemplate: IOutboundProviderTemplate = {
	id: faker.string.uuid(),
	name: faker.lorem.words(3),
	language: 'en',
	type: 'whatsapp',
	category: 'MARKETING',
	status: 'APPROVED',
	qualityScore: {
		score: 'GREEN',
		reasons: null,
	},
	components: [
		{
			type: 'HEADER',
			format: 'TEXT',
			text: 'New {{1}} appointment',
		},
		{
			type: 'BODY',
			text: '**Hi {{1}}** Your _appointment_ for {{2}} is scheduled for {{3}} and can be rescheduled to {{4}} if {{5}} becomes available. {{6}} what do you choose?',
		},
		{
			type: 'FOOTER',
			text: 'Need to reschedule? Tap below to reply',
		},
	],
	createdAt: '', // ISO 8601 timestamp
	createdBy: '',
	modifiedAt: '', // ISO 8601 timestamp
	modifiedBy: '',
	namespace: '',
	wabaAccountId: '',
	// This is the phone number that will be used to send the message.
	phoneNumber: '+5547998461115',
	partnerId: '',
	externalId: '',
	updatedExternal: '', // ISO 8601 timestamp
	rejectedReason: null,
};

export const mockActivityTemplate = {
	template: mockTemplate,
	templateParameters: {
		HEADER: ['John Doe'],
		BODY: ['john.doe@email.com'],
	},
	contactName: 'John Doe',
	providerType: 'phone' as const,
	providerName: 'WhatsApp',
	departmentName: 'Achromatic Ablation',
	agentName: 'Jane Smith',
	username: 'janes.smith',
	sender: '+554766789954',
	recipient: '+554766789959',
};
