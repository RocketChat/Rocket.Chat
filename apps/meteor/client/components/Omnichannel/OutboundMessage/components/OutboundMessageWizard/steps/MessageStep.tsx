import { useEffectEvent } from '@rocket.chat/fuselage-hooks';
import { useWizardContext, WizardActions, WizardBackButton, WizardNextButton } from '@rocket.chat/ui-client';
import type { ComponentProps } from 'react';

import type { MessageFormSubmitPayload } from '../forms/MessageForm/MessageForm';
import MessageForm from '../forms/MessageForm/MessageForm';

type MessageStepProps = Omit<ComponentProps<typeof MessageForm>, 'onSubmit'> & {
	onSubmit(values: MessageFormSubmitPayload): void;
};

const MessageStep = ({ contact, templates, defaultValues, onSubmit }: MessageStepProps) => {
	const { next } = useWizardContext();

	const handleSubmit = useEffectEvent(async (values: MessageFormSubmitPayload) => {
		onSubmit(values);
		next();
	});

	return (
		<div>
			<MessageForm
				contact={contact}
				templates={templates}
				defaultValues={defaultValues}
				onSubmit={handleSubmit}
				renderActions={({ isSubmitting }) => (
					<WizardActions>
						<WizardBackButton />
						<WizardNextButton manual type='submit' loading={isSubmitting} />
					</WizardActions>
				)}
			/>
		</div>
	);
};

export default MessageStep;
