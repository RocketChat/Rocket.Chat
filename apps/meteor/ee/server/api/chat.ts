import type { IMessage, ReadReceipt } from '@rocket.chat/core-typings';
import { License } from '@rocket.chat/license';
import { Meteor } from 'meteor/meteor';

import { API } from '../../../app/api/server/api';
import { getReadReceiptsFunction } from '../methods/getReadReceipts';

type GetMessageReadReceiptsProps = {
	messageId: IMessage['_id'];
};

declare module '@rocket.chat/rest-typings' {
	// eslint-disable-next-line @typescript-eslint/naming-convention
	interface Endpoints {
		'/v1/chat.getMessageReadReceipts': {
			GET: (params: GetMessageReadReceiptsProps) => {
				receipts: ReadReceipt[];
			};
		};
	}
}

API.v1.addRoute(
	'chat.getMessageReadReceipts',
	{ authRequired: true },
	{
		async get() {
			if (!License.hasModule('message-read-receipt')) {
				throw new Meteor.Error('error-action-not-allowed', 'This is an enterprise feature');
			}

			const { messageId } = this.queryParams;
			if (!messageId) {
				return API.v1.failure({
					error: "The required 'messageId' param is missing.",
				});
			}

			return API.v1.success({
				receipts: await getReadReceiptsFunction(messageId, this.userId),
			});
		},
	},
);
