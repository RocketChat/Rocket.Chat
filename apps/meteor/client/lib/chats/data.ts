import type { IEditedMessage, IMessage, IRoom, ISubscription } from '@rocket.chat/core-typings';
import { Random } from '@rocket.chat/random';
import moment from 'moment';

import type { DataAPI } from './ChatAPI';
import { hasAtLeastOnePermission, hasPermission } from '../../../app/authorization/client';
import { Messages, Rooms, Subscriptions } from '../../../app/models/client';
import { settings } from '../../../app/settings/client';
import { MessageTypes } from '../../../app/ui-utils/client';
import { sdk } from '../../../app/utils/client/lib/SDKClient';
import { prependReplies } from '../utils/prependReplies';

export const createDataAPI = ({ rid, tmid }: { rid: IRoom['_id']; tmid: IMessage['_id'] | undefined }): DataAPI => {
	const composeMessage = async (
		text: string,
		{ sendToChannel, quotedMessages, originalMessage }: { sendToChannel?: boolean; quotedMessages: IMessage[]; originalMessage?: IMessage },
	): Promise<IMessage> => {
		const msg = await prependReplies(text, quotedMessages);

		const effectiveRID = originalMessage?.rid ?? rid;
		const effectiveTMID = originalMessage ? originalMessage.tmid : tmid;

		return {
			_id: originalMessage?._id ?? Random.id(),
			rid: effectiveRID,
			...(effectiveTMID && {
				tmid: effectiveTMID,
				...(sendToChannel && { tshow: sendToChannel }),
			}),
			msg,
		} as IMessage;
	};

	const findMessageByID = async (mid: IMessage['_id']): Promise<IMessage | null> =>
		Messages.findOne({ _id: mid, _hidden: { $ne: true } }, { reactive: false }) ?? sdk.call('getSingleMessage', mid);

	const getMessageByID = async (mid: IMessage['_id']): Promise<IMessage> => {
		const message = await findMessageByID(mid);

		if (!message) {
			throw new Error('Message not found');
		}

		return message;
	};

	const findLastMessage = async (): Promise<IMessage | undefined> =>
		Messages.findOne({ rid, tmid: tmid ?? { $exists: false }, _hidden: { $ne: true } }, { sort: { ts: -1 }, reactive: false });

	const getLastMessage = async (): Promise<IMessage> => {
		const message = await findLastMessage();

		if (!message) {
			throw new Error('Message not found');
		}

		return message;
	};

	const findLastOwnMessage = async (): Promise<IMessage | undefined> => {
		const uid = Meteor.userId();

		if (!uid) {
			return undefined;
		}

		return Messages.findOne(
			{ rid, 'tmid': tmid ?? { $exists: false }, 'u._id': uid, '_hidden': { $ne: true } },
			{ sort: { ts: -1 }, reactive: false },
		);
	};

	const getLastOwnMessage = async (): Promise<IMessage> => {
		const message = await findLastOwnMessage();

		if (!message) {
			throw new Error('Message not found');
		}

		return message;
	};

	const canUpdateMessage = async (message: IMessage): Promise<boolean> => {
		if (MessageTypes.isSystemMessage(message)) {
			return false;
		}

		const canEditMessage = hasAtLeastOnePermission('edit-message', message.rid);
		const editAllowed = (settings.get('Message_AllowEditing') as boolean | undefined) ?? false;
		const editOwn = message?.u && message.u._id === Meteor.userId();

		if (!canEditMessage && (!editAllowed || !editOwn)) {
			return false;
		}

		const blockEditInMinutes = settings.get('Message_AllowEditing_BlockEditInMinutes') as number | undefined;
		const bypassBlockTimeLimit = hasPermission('bypass-time-limit-edit-and-delete', message.rid);

		const elapsedMinutes = moment().diff(message.ts, 'minutes');
		if (!bypassBlockTimeLimit && elapsedMinutes && blockEditInMinutes && elapsedMinutes > blockEditInMinutes) {
			return false;
		}

		return true;
	};

	const findPreviousOwnMessage = async (message: IMessage): Promise<IMessage | undefined> => {
		const uid = Meteor.userId();

		if (!uid) {
			return undefined;
		}

		const msg = Messages.findOne(
			{ rid, 'tmid': tmid ?? { $exists: false }, 'u._id': uid, '_hidden': { $ne: true }, 'ts': { $lt: message.ts } },
			{ sort: { ts: -1 }, reactive: false },
		);

		if (!msg) {
			return undefined;
		}

		if (await canUpdateMessage(msg)) {
			return msg;
		}

		return findPreviousOwnMessage(msg);
	};

	const getPreviousOwnMessage = async (message: IMessage): Promise<IMessage> => {
		const previousMessage = await findPreviousOwnMessage(message);

		if (!previousMessage) {
			throw new Error('Message not found');
		}

		return previousMessage;
	};

	const findNextOwnMessage = async (message: IMessage): Promise<IMessage | undefined> => {
		const uid = Meteor.userId();

		if (!uid) {
			return undefined;
		}

		const msg = Messages.findOne(
			{ rid, 'tmid': tmid ?? { $exists: false }, 'u._id': uid, '_hidden': { $ne: true }, 'ts': { $gt: message.ts } },
			{ sort: { ts: 1 }, reactive: false },
		);

		if (!msg) {
			return undefined;
		}

		if (await canUpdateMessage(msg)) {
			return msg;
		}

		return findNextOwnMessage(msg);
	};

	const getNextOwnMessage = async (message: IMessage): Promise<IMessage> => {
		const nextMessage = await findNextOwnMessage(message);

		if (!nextMessage) {
			throw new Error('Message not found');
		}

		return nextMessage;
	};

	const pushEphemeralMessage = async (message: Omit<IMessage, 'rid' | 'tmid'>): Promise<void> => {
		Messages.upsert({ _id: message._id }, { $set: { ...message, rid, ...(tmid && { tmid }) } });
	};

	const updateMessage = async (message: IEditedMessage, previewUrls?: string[]): Promise<void> =>
		sdk.call('updateMessage', message, previewUrls);

	const canDeleteMessage = async (message: IMessage): Promise<boolean> => {
		const uid = Meteor.userId();

		if (!uid) {
			return false;
		}

		if (MessageTypes.isSystemMessage(message)) {
			return false;
		}

		const forceDeleteAllowed = hasPermission('force-delete-message', message.rid);
		if (forceDeleteAllowed) {
			return true;
		}

		const deletionEnabled = settings.get('Message_AllowDeleting') as boolean | undefined;
		if (!deletionEnabled) {
			return false;
		}

		const deleteAnyAllowed = hasPermission('delete-message', rid);
		const deleteOwnAllowed = hasPermission('delete-own-message');
		const deleteAllowed = deleteAnyAllowed || (deleteOwnAllowed && message?.u && message.u._id === Meteor.userId());

		if (!deleteAllowed) {
			return false;
		}

		const blockDeleteInMinutes = settings.get('Message_AllowDeleting_BlockDeleteInMinutes') as number | undefined;
		const bypassBlockTimeLimit = hasPermission('bypass-time-limit-edit-and-delete', message.rid);
		const elapsedMinutes = moment().diff(message.ts, 'minutes');
		const onTimeForDelete = bypassBlockTimeLimit || !blockDeleteInMinutes || !elapsedMinutes || elapsedMinutes <= blockDeleteInMinutes;

		return deleteAllowed && onTimeForDelete;
	};

	const deleteMessage = async (msgIdOrMsg: IMessage | IMessage['_id']): Promise<void> => {
		let msgId: string;
		let roomId: string;
		if (typeof msgIdOrMsg === 'string') {
			msgId = msgIdOrMsg;
			const msg = await findMessageByID(msgId);
			if (!msg) {
				throw new Error('Message not found');
			}
			roomId = msg.rid;
		} else {
			msgId = msgIdOrMsg._id;
			roomId = msgIdOrMsg.rid;
		}

		await sdk.rest.post('/v1/chat.delete', { msgId, roomId });
	};

	const drafts = new Map<IMessage['_id'] | undefined, string>();

	const getDraft = async (mid: IMessage['_id'] | undefined): Promise<string | undefined> => drafts.get(mid);

	const discardDraft = async (mid: IMessage['_id'] | undefined): Promise<void> => {
		drafts.delete(mid);
	};

	const saveDraft = async (mid: IMessage['_id'] | undefined, draft: string): Promise<void> => {
		drafts.set(mid, draft);
	};

	const findRoom = async (): Promise<IRoom | undefined> => Rooms.findOne({ _id: rid }, { reactive: false });

	const getRoom = async (): Promise<IRoom> => {
		const room = await findRoom();

		if (!room) {
			throw new Error('Room not found');
		}

		return room;
	};

	const isSubscribedToRoom = async (): Promise<boolean> => !!Subscriptions.findOne({ rid }, { reactive: false });

	const joinRoom = async (): Promise<void> => {
		await sdk.call('joinRoom', rid);
	};

	const findDiscussionByID = async (drid: IRoom['_id']): Promise<IRoom | undefined> =>
		Rooms.findOne({ _id: drid, prid: { $exists: true } }, { reactive: false });

	const getDiscussionByID = async (drid: IRoom['_id']): Promise<IRoom> => {
		const discussion = await findDiscussionByID(drid);

		if (!discussion) {
			throw new Error('Discussion not found');
		}

		return discussion;
	};

	const createStrictGetter = <TFind extends (...args: any[]) => Promise<any>>(
		find: TFind,
		errorMessage: string,
	): ((...args: Parameters<TFind>) => Promise<Exclude<Awaited<ReturnType<TFind>>, undefined>>) => {
		return async (...args) => {
			const result = await find(...args);

			if (!result) {
				throw new Error(errorMessage);
			}

			return result;
		};
	};

	const findSubscription = async (): Promise<ISubscription | undefined> => {
		return Subscriptions.findOne({ rid }, { reactive: false });
	};

	const getSubscription = createStrictGetter(findSubscription, 'Subscription not found');

	const findSubscriptionFromMessage = async (message: IMessage): Promise<ISubscription | undefined> => {
		return Subscriptions.findOne({ rid: message.rid }, { reactive: false });
	};

	const getSubscriptionFromMessage = createStrictGetter(findSubscriptionFromMessage, 'Subscription not found');

	return {
		composeMessage,
		findMessageByID,
		getMessageByID,
		findLastMessage,
		getLastMessage,
		findLastOwnMessage,
		getLastOwnMessage,
		findPreviousOwnMessage,
		getPreviousOwnMessage,
		findNextOwnMessage,
		getNextOwnMessage,
		pushEphemeralMessage,
		canUpdateMessage,
		updateMessage,
		canDeleteMessage,
		deleteMessage,
		getDraft,
		saveDraft,
		discardDraft,
		findRoom,
		getRoom,
		isSubscribedToRoom,
		joinRoom,
		findDiscussionByID,
		getDiscussionByID,
		findSubscription,
		getSubscription,
		findSubscriptionFromMessage,
		getSubscriptionFromMessage,
	};
};
