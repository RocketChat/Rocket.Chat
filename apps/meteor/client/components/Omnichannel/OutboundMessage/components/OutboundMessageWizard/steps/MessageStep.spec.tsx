import { mockAppRoot } from '@rocket.chat/mock-providers';
import { WizardContext, StepsLinkedList } from '@rocket.chat/ui-client';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import type { ComponentProps } from 'react';
import { forwardRef, useImperativeHandle } from 'react';

import MessageStep from './MessageStep';
import { createFakeContact } from '../../../../../../../tests/mocks/data';
import { createFakeOutboundTemplate } from '../../../../../../../tests/mocks/data/outbound-message';
import type { MessageFormRef } from '../forms/MessageForm';
import type MessageForm from '../forms/MessageForm';

const mockSubmit = jest.fn();

const mockMessageFormRender = jest.fn((_props) => <form name='message-form' />);
jest.mock('../forms/MessageForm', () => ({
	__esModule: true,
	default: forwardRef<MessageFormRef, ComponentProps<typeof MessageForm>>((props, ref) => {
		useImperativeHandle(ref, () => ({ submit: mockSubmit }));
		return mockMessageFormRender(props);
	}),
}));

const steps = new StepsLinkedList([
	{ id: 'test-step-1', title: 'Test Step 1' },
	{ id: 'test-step-2', title: 'Test Step 2' },
	{ id: 'test-step-3', title: 'Test Step 3' },
]);

const mockWizardApi = {
	steps,
	currentStep: steps.head?.next ?? null,
	next: jest.fn(),
	previous: jest.fn(),
	register: jest.fn(),
	goTo: jest.fn(),
	resetNextSteps: jest.fn(),
};

const appRoot = mockAppRoot()
	.withJohnDoe()
	.wrap((children) => {
		return <WizardContext.Provider value={mockWizardApi}>{children}</WizardContext.Provider>;
	});

describe('MessageStep', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should pass accessibility tests', async () => {
		const { container } = render(<MessageStep onSubmit={jest.fn()} />, { wrapper: appRoot.build() });
		const results = await axe(container);
		expect(results).toHaveNoViolations();
	});

	it('should render message form with correct props', () => {
		const defaultValues = { templateId: 'test-template-id' };
		const contact = createFakeContact();
		const templates = [createFakeOutboundTemplate()];

		render(<MessageStep contact={contact} templates={templates} defaultValues={defaultValues} onSubmit={jest.fn()} />, {
			wrapper: appRoot.build(),
		});

		expect(screen.getByRole('form')).toBeInTheDocument();
		expect(mockMessageFormRender).toHaveBeenCalledWith(
			expect.objectContaining({
				defaultValues,
				contact,
				templates,
			}),
		);
	});

	it('should call onSubmit with form values when form submits successfully', async () => {
		const expectedPayload = { templateId: 'test-template-id' };
		mockSubmit.mockResolvedValue(expectedPayload);
		const onSubmit = jest.fn();

		render(<MessageStep onSubmit={onSubmit} />, { wrapper: appRoot.build() });

		await userEvent.click(screen.getByRole('button', { name: 'Next' }));

		await waitFor(() => expect(onSubmit).toHaveBeenCalledWith(expectedPayload));
	});

	it('should not call onSubmit and prevent default when form submission rejects', async () => {
		mockSubmit.mockRejectedValueOnce(new Error('Submission Error'));
		const onSubmit = jest.fn();

		render(<MessageStep onSubmit={onSubmit} />, { wrapper: appRoot.build() });

		const nextButton = screen.getByRole('button', { name: 'Next' });
		const clickEvent = new MouseEvent('click', { bubbles: true, cancelable: true });

		nextButton.dispatchEvent(clickEvent);

		await waitFor(() => expect(onSubmit).not.toHaveBeenCalled());
		await waitFor(() => expect(clickEvent.defaultPrevented).toBeTruthy());
	});

	it('should call previous step when back button is clicked', async () => {
		render(<MessageStep onSubmit={jest.fn()} />, { wrapper: appRoot.build() });

		const backButton = screen.getByRole('button', { name: 'Back' });
		await userEvent.click(backButton);

		await waitFor(() => expect(mockWizardApi.previous).toHaveBeenCalled());
	});

	it('shows a loading state on the button while submit is pending', async () => {
		let resolvePromise: (value: unknown) => void = jest.fn();
		mockSubmit.mockImplementation(
			() =>
				new Promise((resolve) => {
					resolvePromise = resolve;
				}),
		);

		render(<MessageStep onSubmit={jest.fn()} />, { wrapper: appRoot.build() });

		const nextButton = screen.getByRole('button', { name: 'Next' });
		await userEvent.click(nextButton);

		expect(nextButton).toBeDisabled();

		resolvePromise(undefined);

		await waitFor(() => expect(nextButton).not.toBeDisabled());
	});

	it('removes loading state if submit rejects', async () => {
		let rejectPromise: (reason?: any) => void = jest.fn();
		mockSubmit.mockReturnValue(
			new Promise((_, reject) => {
				rejectPromise = reject;
			}),
		);

		render(<MessageStep onSubmit={jest.fn()} />, { wrapper: appRoot.build() });

		const nextButton = screen.getByRole('button', { name: 'Next' });
		await userEvent.click(nextButton);

		expect(nextButton).toBeDisabled();

		rejectPromise(new Error('Failed to submit'));

		await waitFor(() => expect(nextButton).not.toBeDisabled());
	});
});
