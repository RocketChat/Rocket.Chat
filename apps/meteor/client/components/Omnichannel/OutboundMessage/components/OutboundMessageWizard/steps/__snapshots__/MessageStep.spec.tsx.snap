// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MessageStep renders Default without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-162rsse"
    >
      <div>
        <div
          data-testid="message-form"
        >
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-box rcx-box--full steps-wizard-footer rcx-css-181ybla"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-button rcx-button-group__item"
              type="button"
            >
              <span
                class="rcx-button--content"
              >
                Back
              </span>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
              type="submit"
            >
              <span
                class="rcx-button--content"
              >
                Next
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MessageStep renders WithDefaultValues without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-162rsse"
    >
      <div>
        <div
          data-testid="message-form"
        >
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-box rcx-box--full steps-wizard-footer rcx-css-181ybla"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-button rcx-button-group__item"
              type="button"
            >
              <span
                class="rcx-button--content"
              >
                Back
              </span>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
              type="submit"
            >
              <span
                class="rcx-button--content"
              >
                Next
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
